[CI] Isolate centered scalar_u8 exact-allocation gate from package-global mallocs

Parent CI tracker: #3776

Discovered after merging: #3843 / PR #3865

Related completed isolation work: #3848, #3849

## Goal

Keep the centered scalar_u8 scorer's exact zero-allocation contract meaningful and stable in the full collections suite by measuring it in the existing fresh-child allocation harness, while continuing to exercise setup, warm preparation, score correctness, and scratch reuse in the parent process and under the race detector.

## Exact failure

Post-merge `main@d81ae897f3b749caac69685dad49cf26a1e0b183` TreeDB run [29601135696](https://github.com/snissn/gomap/actions/runs/29601135696), attempt 1, Ubuntu job [`87953248798`](https://github.com/snissn/gomap/actions/runs/29601135696/job/87953248798), failed only:

```text
--- FAIL: TestColumnGraphScalarU8CenteredQueryScratch2258 (0.02s)
    column_vector_graph_quantized_asset_test.go:957:
    steady-state centered scalar_u8 prepare+score allocs/run=1 want 0
```

The merged root-publication change does not modify this vector implementation or test. On the exact merged contents, the named test passes locally 100/100 in normal mode and 20/20 under `-race`. Previous exact-SHA PR/proof matrices also passed it. This is consistent with the process-global `testing.AllocsPerRun` contamination already characterized in #3848/#3849, but this particular allocation gate is not yet routed through `enterIsolatedVectorAllocationGate`.

The active main matrix must become terminal before its one classified failed-job rerun is considered. The rerun is evidence only and does not replace this repair.

## Test-first implementation

1. Preserve all existing setup, centered-query parity, scratch reuse, warm preparation, and warm scoring assertions in the ordinary parent test.
2. Under `-race`, stop only before `testing.AllocsPerRun`; do not skip the setup/warm correctness path.
3. In normal mode, route the exact allocation measurement through the existing fresh-copy test-binary helper in `TreeDB/collections/vector_allocation_isolation_test.go` with a unique selector.
4. Keep the exact `allocs/run == 0` threshold and 1,000-run sample unchanged in the selected child.
5. Add product code only if the isolated child deterministically allocates; otherwise this remains a test-contract repair.

## Boundaries

- Do not weaken the zero-allocation threshold.
- Do not convert exact allocation evidence into a loose budget or benchmark-only observation.
- Do not skip setup, warm preparation, score parity, or scratch validation under race.
- Do not serialize or disable unrelated collections tests.
- Do not change scalar_u8 scoring behavior unless isolated evidence proves a product regression.

## Performance classification

Performance-contract preserving and expected to be test-only. The gate must continue to prove zero steady-state allocations in a fresh normal-mode child process; the full-package parent proves integration correctness, not process-global malloc accounting.

## Required evidence

- [ ] Pre-fix failure is recorded from exact-main hosted CI.
- [ ] Parent-path named test passes repeatedly in normal and race modes.
- [ ] Selected isolated child passes repeatedly with the unchanged zero-allocation threshold.
- [ ] Full `TreeDB/collections` normal and race suites pass.
- [ ] Exact-head TreeDB actual plus two independent exact-SHA proof matrices pass without using a rerun as primary evidence.
- [ ] Exact-head Codex review is clean and all review threads are resolved.
- [ ] Exact default-branch TreeDB CI is green after merge.

## Branch and PR process

- Start from the exact current `main` after the active post-merge matrix is terminal.
- Keep the patch limited to the centered-query test and the existing isolation helper only if a selector-level helper adjustment is required.
- Publish the exact local commands, hosted run IDs, review artifact, and merge commit.

## Graph role

This node restores a trustworthy green default branch before #3864 begins. It precedes #3864 and final certification #3777. It is independent of #3863's performance-runner variability and #3866's GC fixture settlement.

--- a/.github/scripts/test_treedb_windows_core_headroom.py
+++ b/.github/scripts/test_treedb_windows_core_headroom.py
@@ -83,6 +83,41 @@ class TreeDBWindowsCoreHeadroomTest(unittest.TestCase):
                     rf'\"\${tests_file}\"',
                 )
 
+    def test_collections_runnables_are_split_across_core_shards(self) -> None:
+        workflow = WORKFLOW.read_text(encoding="utf-8")
+        test_job = workflow_job(workflow, "test")
+        core_case = re.search(
+            r"^            windows-core\)\n(?P<body>.*?)(?=^              ;;$)",
+            test_job,
+            re.MULTILINE | re.DOTALL,
+        )
+        self.assertIsNotNone(core_case, "missing windows-core shell case")
+        body = core_case.group("body")
+
+        self.assertIn(
+            "grep -v '^github.com/snissn/gomap/TreeDB/collections$'",
+            body,
+            "collections must not remain an indivisible package-modulo assignment",
+        )
+        self.assertIn(
+            r"go test ./collections -list '^(Test|Example|Fuzz).*' | tr -d '\r' | "
+            r"grep -E '^(Test|Example|Fuzz)' > "
+            '"$collections_all_file"',
+            body,
+            "collections enumeration must retain ordinary fuzz-seed coverage",
+        )
+        self.assertRegex(
+            body,
+            r'awk -v idx="\$package_shard_index" -v n="\$package_shard_count" .* '
+            r'"\$collections_all_file" > "\$collections_test_file"',
+        )
+        self.assertIn(
+            'run_named_test_file ./collections "TreeDB/collections test '
+            'shard=${package_shard_index}/${package_shard_count}" '
+            '"$collections_test_file" treedb-test.jsonl',
+            body,
+        )
+
     def test_caching_shards_keep_their_existing_bounded_caps(self) -> None:
         workflow = WORKFLOW.read_text(encoding="utf-8")
 

--- a/.github/workflows/treedb-tests.yml
+++ b/.github/workflows/treedb-tests.yml
@@ -329,11 +329,15 @@ jobs:
               root_test_file="$(make_tmp)"
               db_all_file="$(make_tmp)"
               db_test_file="$(make_tmp)"
-              go list ./... | grep -v '^github.com/snissn/gomap/TreeDB$' | grep -v '^github.com/snissn/gomap/TreeDB/caching$' | grep -v '^github.com/snissn/gomap/TreeDB/db$' | awk -v idx="$package_shard_index" -v n="$package_shard_count" 'BEGIN { c = 0 } { if ((c % n) == idx) print; c++ }' > "$package_file"
+              collections_all_file="$(make_tmp)"
+              collections_test_file="$(make_tmp)"
+              go list ./... | grep -v '^github.com/snissn/gomap/TreeDB$' | grep -v '^github.com/snissn/gomap/TreeDB/caching$' | grep -v '^github.com/snissn/gomap/TreeDB/collections$' | grep -v '^github.com/snissn/gomap/TreeDB/db$' | awk -v idx="$package_shard_index" -v n="$package_shard_count" 'BEGIN { c = 0 } { if ((c % n) == idx) print; c++ }' > "$package_file"
               go test . -list '^(Test|Example).*' | tr -d '\r' | grep -E '^(Test|Example)' > "$root_all_file"
               go test ./db -list 'Test.*' | tr -d '\r' | grep '^Test' > "$db_all_file"
+              go test ./collections -list '^(Test|Example|Fuzz).*' | tr -d '\r' | grep -E '^(Test|Example|Fuzz)' > "$collections_all_file"
               awk -v idx="$package_shard_index" -v n="$package_shard_count" 'BEGIN { c = 0 } { if ((c % n) == idx) print; c++ }' "$root_all_file" > "$root_test_file"
               awk -v idx="$package_shard_index" -v n="$package_shard_count" 'BEGIN { c = 0 } { if ((c % n) == idx) print; c++ }' "$db_all_file" > "$db_test_file"
+              awk -v idx="$package_shard_index" -v n="$package_shard_count" 'BEGIN { c = 0 } { if ((c % n) == idx) print; c++ }' "$collections_all_file" > "$collections_test_file"
               echo "TreeDB windows-core package shard=${package_shard_index}/${package_shard_count}"
               ran_tests=false
               if [ -s "$package_file" ]; then
@@ -346,6 +350,10 @@ jobs:
               fi
               if [ -s "$db_test_file" ]; then
                 run_named_test_file ./db "TreeDB/db test shard=${package_shard_index}/${package_shard_count}" "$db_test_file" treedb-test.jsonl
+                ran_tests=true
+              fi
+              if [ -s "$collections_test_file" ]; then
+                run_named_test_file ./collections "TreeDB/collections test shard=${package_shard_index}/${package_shard_count}" "$collections_test_file" treedb-test.jsonl
                 ran_tests=true
               fi
               if [ "$ran_tests" = false ]; then

--- a/TreeDB/collections/column_vector_graph_quantized_asset_test.go
+++ b/TreeDB/collections/column_vector_graph_quantized_asset_test.go
@@ -941,6 +941,12 @@ func TestColumnGraphScalarU8CenteredQueryScratch2258(t *testing.T) {
 	if _, err := scorer.scoreOrdinal(1, &scratch, nil); err != nil {
 		t.Fatalf("warm scoreOrdinal: %v", err)
 	}
+	if collectionsRaceEnabled {
+		t.Skip("exact allocation counts are unstable under race instrumentation")
+	}
+	if !enterIsolatedVectorAllocationGate(t, "scalar-u8-centered-query-scratch") {
+		return
+	}
 	var stats columnVectorGraphNativeSearchStats
 	allocs := testing.AllocsPerRun(1000, func() {
 		scorer, err := reader.prepareScalarU8QuantizedScorer(columnVectorGraphNativeSearchQueryModeQuantizedOnly, def.QuantizedIndexes[0].Name, query, queryInvNorm, &scratch)

